### PR TITLE
Mypy disallow untyped calls

### PR DESCRIPTION
**Pull request by zundertj**
See #4053 for the original attempt. That one got merged and reverted, so this is the new one.

This PR:
* sets mypy config `disallow_untyped_calls = true`
* adjusts the type annotations accordingly
* runs into an issue where numpy < 1.22 does not have particular type annotations, and numpy 1.22+ does have
* as CI tests Python 3.7 and 3.10, and the latest numpy supporting Python 3.7 is 1.21.6, this PR ends up in a situation where it cannot please mypy on both Python versions (unless we would force an old numpy, but that is not future proof)
* therefore it turns off the flag in CI only! for Python 3.7 (beware if you are devving locally with an old numpy version)
* GH actions has nice support for matrix, but does not allow if-statements in step definitions based on the matrix version
* thus we are forced to copy-paste the Python test job for Python 3.7, with the only difference being the flag

### CODE REVIEW

File: .github/workflows/test-python.yaml [hunk 1]
```diff
@@ -50,3 +50,49 @@ jobs:
         run: |
           pip uninstall pandas -y
           python -c "import polars"
+
```
Comment (ritchie46) [inline .github/workflows/test-python.yaml:53]: Can you add a trigger that on this snippet that will also run when this yaml is changed?

```
on:
  pull_request:
    paths:
      - 'py-polars/**'
- 'polars/**'

```

Comment (zundertj) [reply .github/workflows/test-python.yaml:53]: Maybe I am misunderstanding, but the same `on` clause is currently defined on lines 3 to 7, outside the `jobs`, so I assume it works for both? See also https://github.com/pola-rs/polars/runs/7436128231?check_suite_focus=true for the py3.7 run.

Comment (ritchie46) [reply .github/workflows/test-python.yaml:53]: No, I mean that we also trigger the job when the job definition itself has changed. Currently it is only triggered when python code has changed.

Comment (zundertj) [reply .github/workflows/test-python.yaml:53]: Let me raise a separate PR for that.

Comment (zundertj) [reply .github/workflows/test-python.yaml:53]: See #4125

**Comment by ritchie46** [comment]
Can you do a rebase now that #4125 is merged?

**Comment by zundertj** [comment]
Coverage job hangs, but otherwise it works.

**Pull request merged**